src/utils/keyRotation.ts

```diff
@@ -1,17 +1,43 @@
 import { Bindings } from '../bindings';
 import cronParser from 'cron-parser';
 
-export function shouldRotateKey(date: Date, env: Bindings): boolean {
-	const utcDate = new Date(date.toISOString());
+interface CronParseResult {
+	prevTime?: number;
+	nextTime?: number;
+	match: boolean;
+}
+
+const KEY_LIFESPAN = 48 * 60 * 60 * 1000;
 
+export function getPrevRotationTime(mostRecentKeyUploadTime: Date, env: Bindings): number | null {
+	let effectivePrevTime: number;
 	if (env.ROTATION_CRON_STRING) {
-		const result = matchCronTime(env.ROTATION_CRON_STRING, utcDate);
-		return result;
+		const { prevTime } = matchCronTime(env.ROTATION_CRON_STRING, mostRecentKeyUploadTime);
+
+		if (prevTime === undefined) {
+			console.error('Failed to determine previous rotation time for key');
+			return null;
+		}
+
+		effectivePrevTime = Math.max(prevTime, mostRecentKeyUploadTime.getTime());
+	} else {
+		effectivePrevTime = mostRecentKeyUploadTime.getTime();
 	}
-	return false;
+	return effectivePrevTime;
+}
+
+export function shouldRotateKey(date: Date, env: Bindings): boolean {
+	const utcDate = new Date(date.toISOString());
+	return env.ROTATION_CRON_STRING ? matchCronTime(env.ROTATION_CRON_STRING, utcDate).match : false;
 }
 
-function matchCronTime(cronString: string, date: Date): boolean {
+export function shouldClearKey(keyUploadTime: Date, now: Date, effectivePrevTime: number): boolean {
```

Comment:
As long as that constants like KEY_LIFESPAN and variables like effectivePrevTime are always well-defined, this comparisons seem correct.